### PR TITLE
[FLINK-8975] [test] Add resume from savepoint end-to-end test

### PR DESCRIPTION
**Pull request by tzulitai**
## What is the purpose of the change

This pull request adds an end-to-end test that verifies resuming a job from a savepoint.

The complete end-to-end test consists of the following:

1. The `StateMachineExample` is used for the end-to-end test
2. A separate job to generate Kafka events for the state machine is run
3. After the state machine job runs for a while, we take a savepoint.
4. Cancelling and resuming the state machine example job with the savepoint.

All the above steps should not result in any errors or outputs from the state machine job. If so, the end-to-end test would fail.

## Brief change log

- Add a separate main class for the Kafka events generator job
- Add `test_resume_savepoint.sh` test script

## Verifying this change

This PR itself introduces a new test.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-end-to-end-tests/test-scripts/test_resume_savepoint.sh [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+# then, run the events generator
+EVENTS_GEN_JOB=$($FLINK_DIR/bin/flink run -d -c org.apache.flink.streaming.examples.statemachine.KafkaEventsGeneratorJob $FLINK_DIR/examples/streaming/StateMachineExample.jar \
+  --kafka-topic test-input --sleep 200 \
+  | grep "Job has been submitted with JobID" | sed 's/.* //g')
+
+wait_job_running $EVENTS_GEN_JOB
+
+# wait a bit to have some events pass through the state machine
+sleep 15
```
Comment (tzulitai) [inline flink-end-to-end-tests/test-scripts/test_resume_savepoint.sh]: An alternative to this:
Query "read-records" metrics via the REST API, and only proceed after the job has processed a said number of records.
This however requires adding an extra dependency, such as jq, for the response parsing.

File: flink-end-to-end-tests/test-scripts/test_resume_savepoint.sh [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+# resume state machine job with savepoint
+STATE_MACHINE_JOB=$($FLINK_DIR/bin/flink run -s $SAVEPOINT_PATH -d $FLINK_DIR/examples/streaming/StateMachineExample.jar \
+  --kafka-topic test-input \
+  | grep "Job has been submitted with JobID" | sed 's/.* //g')
+
+wait_job_running $STATE_MACHINE_JOB
+
+sleep 15
+
+# if state is errorneous and the state machine job produces alerting state transitions,
```
Comment (zentol) [inline flink-end-to-end-tests/test-scripts/test_resume_savepoint.sh]: when do the jobs shut down if no error occurs?

File: flink-end-to-end-tests/test-scripts/test_resume_savepoint.sh (hunk 1 above)
Comment (zentol) [reply flink-end-to-end-tests/test-scripts/test_resume_savepoint.sh]: you could use the log4j reporter and grep the logs.

File: flink-end-to-end-tests/test-scripts/test_resume_savepoint.sh (hunk 2 above)
Comment (tzulitai) [reply flink-end-to-end-tests/test-scripts/test_resume_savepoint.sh]: when the script exits, a cleanup hook shuts down the cluster. It also parses the logs for any unexpected errors; if there is one, the test fails.

File: flink-end-to-end-tests/test-scripts/test_resume_savepoint.sh (hunk 1 above)
Comment (tzulitai) [reply flink-end-to-end-tests/test-scripts/test_resume_savepoint.sh]: That's a good idea! I'll give this approach a try.

File: flink-end-to-end-tests/test-scripts/test_resume_savepoint.sh (hunk 2 above)
Comment (zentol) [reply flink-end-to-end-tests/test-scripts/test_resume_savepoint.sh]: The only trap is see is `kafka_cleanup`, which shuts down ZK and kafka. What about the state machine though?

File: flink-end-to-end-tests/test-scripts/test_resume_savepoint.sh [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+$KAFKA_DIR/bin/kafka-server-start.sh -daemon $KAFKA_DIR/config/server.properties
+
+# make sure to stop Kafka and ZooKeeper at the end
+
+function kafka_cleanup {
+  $KAFKA_DIR/bin/kafka-server-stop.sh
+  $KAFKA_DIR/bin/zookeeper-server-stop.sh
+
+  # make sure to run regular cleanup as well
+  cleanup
```
Comment (tzulitai) [inline flink-end-to-end-tests/test-scripts/test_resume_savepoint.sh]: @zentol the `kafka_cleanup` trap also includes this, which shuts down the Flink cluster and checks logs for errors.

**Comment by tzulitai** [comment]
@zentol I've updated the e2e test to use the SLF4J reporter to monitor progress, instead of naive sleeping. Could you have another look?

File: flink-end-to-end-tests/test-scripts/test_resume_savepoint.sh [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+source "$(dirname "$0")"/common.sh
+
+# modify configuration to have 2 slots
+cp $FLINK_DIR/conf/flink-conf.yaml $FLINK_DIR/conf/flink-conf.yaml.bak
+sed -i -e 's/taskmanager.numberOfTaskSlots: 1/taskmanager.numberOfTaskSlots: 2/' $FLINK_DIR/conf/flink-conf.yaml
+
+# modify configuration to use SLF4J reporter; we will be using this to monitor the state machine progress
+cp $FLINK_DIR/opt/flink-metrics-slf4j-1.6-SNAPSHOT.jar $FLINK_DIR/lib/
+echo "metrics.reporter.slf4j.class: org.apache.flink.metrics.slf4j.Slf4jReporter" >> $FLINK_DIR/conf/flink-conf.yaml
+echo "metrics.reporter.slf4j.interval: 5 SECONDS" >> $FLINK_DIR/conf/flink-conf.yaml
```
Comment (zentol) [inline flink-end-to-end-tests/test-scripts/test_resume_savepoint.sh]: that's a pretty large report interval which probably makes the test take longer than necessary.

File: flink-end-to-end-tests/test-scripts/test_resume_savepoint.sh [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+start_cluster
+
+# get Kafka 0.10.0
```
Comment (zentol) [inline flink-end-to-end-tests/test-scripts/test_resume_savepoint.sh]: I would do this before starting the flink cluster

File: flink-end-to-end-tests/test-scripts/test_resume_savepoint.sh [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+  fi
+done
+
+# take a savepoint of the state machine job
+SAVEPOINT_PATH=$(take_savepoint $STATE_MACHINE_JOB $TEST_DATA_DIR \
+  | grep "Savepoint completed. Path:" | sed 's/.* //g')
+
+cancel_job $STATE_MACHINE_JOB
+
+# Since it is not possible to differentiate reporter output between the first and second exuection,
```
Comment (zentol) [inline flink-end-to-end-tests/test-scripts/test_resume_savepoint.sh]: typo: execution

**Pull request closed without merging**